ompi/mca/coll/base/coll_base_allreduce.c

```diff
@@ -367,9 +367,22 @@ ompi_coll_base_allreduce_intra_ring(const void *sbuf, void *rbuf, int count,
         return MPI_SUCCESS;
     }
 
+    /* Special case for non-commutative ops - use recursive doubling */
+    if (!ompi_op_is_commute(op)) {
```

Comment:
I would have moved this check into a common location, but then if the user calls a specific collective algorithm directly it would not work. That means we need to have this check in every single function that can be selected by the users.